Add empty_generic, which allocates an empty tensor in a device-generic way

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Restore storage on meta tensors; increase meta coverage
* #54711 Compute Tensor::toString() without reference to backend
* #54710 Make memory overlap a little less precise so it works with null data ptr
* #54709 Make the fuser raise NotImplementedError when unknown device is hit
* #54708 Avoid testing device in cdist when called in a "Math" context
* #54707 Make convolution_overrideable default implementation raise NotImplementedError
* #54706 Disambiguate where 'Doesn't run' error message comes from
* #54705 Fix missing format string qualifier
* #54704 Rename linalg.det test so that it generates a valid method name
* **#54703 Add empty_generic, which allocates an empty tensor in a device-generic way**
* #54597 Fix some more native_functions.yaml mistakes
* #54627 Nasty little hack to preserve NotImplementedError raised in interpreter

The trick is that this function takes in the allocator and dispatch key
explicitly; so you still need to know where to find the appropriate
allocator.  The plan is to use this for meta tensors, but you probably
could also use this for empty_cuda as well.  It also takes in arguments
post optional resolution, which can save a few instructions if you want
to call this function directly (no uses yet).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27338814](https://our.internmc.facebook.com/intern/diff/D27338814)

ezyang committed:
Add empty_generic, which allocates an empty tensor in a device-generic way

The trick is that this function takes in the allocator and dispatch key
explicitly; so you still need to know where to find the appropriate
allocator.  The plan is to use this for meta tensors, but you probably
could also use this for empty_cuda as well.  It also takes in arguments
post optional resolution, which can save a few instructions if you want
to call this function directly (no uses yet).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

iseeyuan reviewed (approved):
It's nice to have empty_generic function!

ezyang committed:
Update on "Add empty_generic, which allocates an empty tensor in a device-generic way"


The trick is that this function takes in the allocator and dispatch key
explicitly; so you still need to know where to find the appropriate
allocator.  The plan is to use this for meta tensors, but you probably
could also use this for empty_cuda as well.  It also takes in arguments
post optional resolution, which can save a few instructions if you want
to call this function directly (no uses yet).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27338814](https://our.internmc.facebook.com/intern/diff/D27338814)

[ghstack-poisoned]
